### PR TITLE
importinto: calculate writer mem size using step resource & add a option for test

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #49008

Problem Summary:

### What changed and how does it work?
- calculate writer mem size using step resource & add a test option to force merge-step
- add some logs
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
before this pr, import 1t data
```
2024-03-05 18:21:09 [task_executor.go:309] ["execute task step completed"] [task-id=4] [task-type=ImportInto] [step=encode&sort] [mem-limit-percent=0.99] [server-mem-limit=88%] [resource="[CPU=16, Mem=27.83GiB]"] [takeTime=1h34m2.774450553s] []
2024-03-05 19:04:51 [task_executor.go:309] ["execute task step completed"] [task-id=4] [task-type=ImportInto] [step=write&ingest] [mem-limit-percent=0.99] [server-mem-limit=88%] [resource="[CPU=16, Mem=27.83GiB]"] [takeTime=42m49.945995706s] []
2024-03-05 19:13:38 [task_executor.go:309] ["execute task step completed"] [task-id=4] [task-type=ImportInto] [step=post-process] [mem-limit-percent=0.99] [server-mem-limit=88%] [resource="[CPU=16, Mem=27.83GiB]"] [takeTime=8m46.067248935s] []
```
after, the time to encode and ingest is similar, the checksum time increase, but it's not related to this pr change
```
2024-03-05 21:41:52 [task_executor.go:309] ["execute task step completed"] [task-id=1] [task-type=ImportInto] [step=encode&sort] [mem-limit-percent=0.99] [server-mem-limit=88%] [resource="[CPU=16, Mem=27.83GiB]"] [takeTime=1h33m19.647405717s] []
2024-03-05 22:25:21 [task_executor.go:309] ["execute task step completed"] [task-id=1] [task-type=ImportInto] [step=write&ingest] [mem-limit-percent=0.99] [server-mem-limit=88%] [resource="[CPU=16, Mem=27.83GiB]"] [takeTime=42m5.874066197s] []
2024-03-05 22:53:11 [task_executor.go:309] ["execute task step completed"] [task-id=1] [task-type=ImportInto] [step=post-process] [mem-limit-percent=0.99] [server-mem-limit=88%] [resource="[CPU=16, Mem=27.83GiB]"] [takeTime=27m49.204570776s] []
```

with this pr, run import 1t with merge-step, success
```
[task_executor.go:309] ["execute task step completed"] [task-id=1] [task-type=ImportInto] [step=merge-sort] [mem-limit-percent=0.99] [server-mem-limit=88%] [resource="[CPU=16, Mem=27.83GiB]"] [takeTime=31m43.024898826s] []
```
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by D3Hunter** [comment]
@lance6716

File: br/pkg/lightning/backend/external/merge.go [hunk 1]
```diff
@@ -76,6 +76,22 @@ func MergeOverlappingFiles(
...
 // mergeOverlappingFilesInternal reads from given files whose key range may overlap
 // and writes to one new sorted, nonoverlapping files.
+// since some memory are taken by library, such as HTTP2, that we cannot calculate
+// accurately, here we only consider the memory used by our code, the estimate max
+// memory usage of this function is:
+//
+//	memSizeLimit
+//	+ 20 * partSize
+//	+ 20 * 5MiB(stat file, we might not use all part, as stat file is quite small)
+//	+ readBufferSize * len(paths)
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/merge.go:86]: in NewMergeKVIter, there's an extra memory consumption of  concurrenct reading

Comment (D3Hunter) [reply br/pkg/lightning/backend/external/merge.go:86]: https://github.com/pingcap/tidb/pull/51486/commits/825d84cb8b745056578c7e8a81eb9202d16725c8

**Comment by D3Hunter** [comment]
update result of merge-sort
@lance6716

**Comment by D3Hunter** [comment]
@lance6716

File: pkg/disttask/importinto/planner.go [hunk 2]
```diff
@@ -482,7 +482,7 @@ func getSortedKVMetasForIngest(planCtx planner.PlanCtx) (map[string]*external.So
 	for kvGroup, kvMeta := range kvMetasOfEncodeStep {
 		// only part of kv files are merge sorted. we need to merge kv metas that
 		// are not merged into the kvMetasOfMergeSort.
-		if skipMergeSort(kvGroup, kvMeta.MultipleFilesStats) {
+		if !p.Plan.ForceMergeStep && skipMergeSort(kvGroup, kvMeta.MultipleFilesStats) {
```
Comment (lance6716) [inline pkg/disttask/importinto/planner.go:485]: maybe we can remove the failpoint, `ForceMergeStep` seems cover all calling

**Review by lance6716 — APPROVED** [review]

File: pkg/disttask/importinto/planner.go (hunk 2 above)
Comment (D3Hunter) [reply pkg/disttask/importinto/planner.go:485]: i deleted it once, but find that some test only want to skip for some kv group

**Pull request merged**